crates/red_knot_test/src/parser.rs

```diff
@@ -0,0 +1,467 @@
+
+pub(crate) fn parse<'s>(path: &'s str, source: &'s str) -> anyhow::Result<MarkdownTestSuite<'s>> {
+    let parser = Parser::new(path, source);
+    parser.parse()
+}
+
+#[newtype_index]
+struct SectionId;
+
+#[newtype_index]
+struct FileId;
+
+/// A parsed markdown file containing tests.
+///
+/// Borrows from the source string and filepath it was created from.
+#[derive(Debug)]
+pub(crate) struct MarkdownTestSuite<'s> {
+    /// Header sections.
+    sections: IndexVec<SectionId, Section<'s>>,
+
+    /// Test files embedded within the Markdown file.
+    files: IndexVec<FileId, File<'s>>,
+}
+
+impl<'s> MarkdownTestSuite<'s> {
+    pub(crate) fn tests<'m>(&'m self) -> MarkdownTestIterator<'m, 's> {
+        MarkdownTestIterator {
+            suite: self,
+            current_file_index: 0,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct MarkdownTest<'m, 's> {
+    suite: &'m MarkdownTestSuite<'s>,
+    section: &'m Section<'s>,
+    files: &'m [File<'s>],
+}
+
+impl<'m, 's> MarkdownTest<'m, 's> {
+    pub(crate) fn name(&self) -> String {
+        let mut name = self.section.title.to_string();
+        let mut parent_id = self.section.parent_id;
+        while let Some(parent) = parent_id.map(|section_id| &self.suite.sections[section_id]) {
+            parent_id = parent.parent_id;
+            name.insert_str(0, " - ");
+            name.insert_str(0, parent.title);
+        }
+        name
+    }
+
+    pub(crate) fn files(&self) -> impl Iterator<Item = &'m File<'s>> {
+        self.files.iter()
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct MarkdownTestIterator<'m, 's> {
+    suite: &'m MarkdownTestSuite<'s>,
+    current_file_index: usize,
+}
+
+impl<'m, 's> Iterator for MarkdownTestIterator<'m, 's> {
+    type Item = MarkdownTest<'m, 's>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut current_file_index = self.current_file_index;
+        let mut file = self.suite.files.get(current_file_index.into());
+        let section_id = file?.section;
+        while file.is_some_and(|file| file.section == section_id) {
+            current_file_index += 1;
+            file = self.suite.files.get(current_file_index.into());
+        }
+        let files = &self.suite.files
+            [FileId::from_usize(self.current_file_index)..FileId::from_usize(current_file_index)];
+        self.current_file_index = current_file_index;
+        Some(MarkdownTest {
+            suite: self.suite,
+            section: &self.suite.sections[section_id],
+            files,
+        })
+    }
+}
+
+#[derive(Debug)]
+struct Section<'s> {
+    title: &'s str,
+    level: usize,
+    parent_id: Option<SectionId>,
+}
+
+#[derive(Debug)]
+pub(crate) struct File<'s> {
+    section: SectionId,
+    pub(crate) path: &'s str,
+    pub(crate) lang: &'s str,
+    pub(crate) code: &'s str,
+}
+
+static HEADER_RE: Lazy<Regex> =
+    Lazy::new(|| Regex::new(r"^(\s*\n)*(?<level>#+)\s+(?<title>.+)\s*\n").unwrap());
```

Comment:
I don't think "greedy" parsing using the regex multiline option works, at least not cleanly, because it means whichever regex I check first (header or code) will just happily bypass/ignore the other one. We don't ever know whether the next item we expect is a header or a code block.